### PR TITLE
fix(tracing): ensure Context is serializable

### LINKED ISSUE #4434: "Tracing Across Processes" example crashes
### Summary of problem

Hello, I cannot get "tracing across processes" to work. Python crashes while trying to pickle the tracer context RLock. The minimal reproduction example is [the example from the ddtrace advanced usage docs](https://ddtrace.readthedocs.io/en/stable/advanced_usage.html#tracing-across-processes). I also cannot get this to work with any prior ddtrace version.

### Which version of dd-trace-py are you using?

ddtrace: 1.5.3
python: 3.9.13

### How can we reproduce your problem?

Run the `Tracing Across Processes` example [from the docs](https://ddtrace.readthedocs.io/en/stable/advanced_usage.html#tracing-across-processes)

```python
from multiprocessing import Process
import time
from ddtrace import tracer

def _target(ctx):
    tracer.context_provider.activate(ctx)
    with tracer.trace("proc"):
        time.sleep(1)
    tracer.shutdown()

with tracer.trace("work"):
    proc = Process(target=_target, args=(tracer.current_trace_context(),))
    proc.start()
    time.sleep(1)
    proc.join()
```

### What is the result that you get?

```
Traceback (most recent call last):
  File "/Users/jboulanger/Library/Application Support/JetBrains/PyCharm2021.2/scratches/scratch.py", line 15, in <module>
    proc.start()
  File "/usr/local/Cellar/python@3.9/3.9.13_3/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/usr/local/Cellar/python@3.9/3.9.13_3/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/usr/local/Cellar/python@3.9/3.9.13_3/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/usr/local/Cellar/python@3.9/3.9.13_3/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/usr/local/Cellar/python@3.9/3.9.13_3/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/usr/local/Cellar/python@3.9/3.9.13_3/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/usr/local/Cellar/python@3.9/3.9.13_3/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
TypeError: cannot pickle '_thread.RLock' object
Process finished with exit code 1
```

### What is the result that you expected?

I expected this to complete without crashing

### PR DESCRIPTION
**Pull request by brettlangdon**
## Description
`ddtrace.context.Context` object is not serializable, meaning it cannot be pickled/shared between processes.

This breaks the example usage we have in our documentation for passing context through to other threads.

e.g. `Process(target=_target, args=(ctx, ))`

This fix added `__getstate__` and `__setstate__` methods to `Context` class to have pickle ignore the RLock which cannot be serialized.

## Checklist
- [x] Add additional sections for `feat` and `fix` pull requests.
- [x] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.

## Relevant issue(s)

Fixes #4434

## Testing strategy

Test cases added for pickling/unpickling `Context` instances, as well as a snapshot test added for the multiprocessing strategy outlined in our documentation.

https://ddtrace.readthedocs.io/en/stable/advanced_usage.html#tracing-across-processes

## Reviewer Checklist
- [x] Title is accurate.
- [x] Description motivates each change.
- [x] No unnecessary changes were introduced in this PR.
- [x] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [x] Release note has been added for fixes and features, or else `changelog/no-changelog` label added.
- [x] All relevant GitHub issues are correctly linked.
- [x] Backports are identified and tagged with Mergifyio.

### CODE REVIEW

File: ddtrace/context.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 from typing import Optional
 from typing import TYPE_CHECKING
 from typing import Text
+from typing import Tuple
```
Comment (mabdinur) [inline ddtrace/context.py:7]: nit: we can move this line under the `if TYPE_CHECKING:` check

**Review by Kyle-Verhoog — APPROVED** [review]

**Comment by Kyle-Verhoog** [comment]
@mergifyio backport 1.5 1.6

File: releasenotes/notes/fix-context-serializable-2c9768cbe4738b73.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+---
+fixes:
+  - |
+    tracing: fix ``ddtrace.context.Context`` class not being serializable by ``pickle`` for passing context across processes.
```
Comment (mabdinur) [inline releasenotes/notes/fix-context-serializable-2c9768cbe4738b73.yaml:4]: The release note is a bit wordy. Can we simplify it a bit? 

```suggestion
    tracing: fix manual tracing across processes, make``ddtrace.context.Context`` pickable.
```

**Review by mabdinur — COMMENTED** [review]
A few nits. Otherwise LGTM

File: ddtrace/context.py (hunk 1 above)
Comment (brettlangdon) [reply ddtrace/context.py:7]: oh sure

File: releasenotes/notes/fix-context-serializable-2c9768cbe4738b73.yaml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+---
+fixes:
+  - |
+    tracing: fix manual tracing across processes, make``ddtrace.context.Context`` pickable.
```
Comment (mabdinur) [inline releasenotes/notes/fix-context-serializable-2c9768cbe4738b73.yaml:4]:
```suggestion
    tracing: fix manual tracing across processes, make ``ddtrace.context.Context`` pickable.
```

**Review by mabdinur — DISMISSED** [review]

File: releasenotes/notes/fix-context-serializable-2c9768cbe4738b73.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+---
+fixes:
+  - |
+    tracing: fix manual tracing across processes, make ``ddtrace.context.Context`` pickable.
```
Comment (brettlangdon) [inline releasenotes/notes/fix-context-serializable-2c9768cbe4738b73.yaml:4]:
```suggestion
    tracing: fix manual tracing across processes, make ``ddtrace.context.Context`` serializable.
```

"pickable" (should probably be "pickleable"?) isn't a real word.

**Review by ZStriker19 — APPROVED** [review]
today I learned

File: releasenotes/notes/fix-context-serializable-2c9768cbe4738b73.yaml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+---
+fixes:
+  - |
+    tracing: fix manual tracing across processes, make ``ddtrace.context.Context`` serializable.
```
Comment (Kyle-Verhoog) [inline releasenotes/notes/fix-context-serializable-2c9768cbe4738b73.yaml:4]:
```suggestion
    tracing: make ``ddtrace.context.Context`` serializable which fixes distributed tracing across processes.
```

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by mabdinur — APPROVED** [review]
